workloads/imagenet/imagenet_pytorch/workload.py

```diff
@@ -0,0 +1,235 @@
+      logits, _ = self.model_fn(
+          params,
+          images,
+          model_state,
+          spec.ForwardPassMode.EVAL,
+          model_rng,
+          update_batch_norm=False)
+      # TODO(znado): add additional eval metrics?
+      batch_metrics = self._eval_metric(logits, labels)
+      total_metrics = {
+          k: v + batch_metrics[k] for k, v in total_metrics.items()
+      }
+      num_batches += 1
+    # Note (runame): the returned result is not exact because the last batch
+    # has a smaller batch size
+    return {k: float(v / num_batches) for k, v in total_metrics.items()}
+
+  def _build_dataset(self,
+      data_rng: spec.RandomState,
+      split: str,
+      data_dir: str,
+      batch_size: int):
+
+    is_train = (split == "train")
+
+    normalize = transforms.Compose([
+        transforms.ToTensor(),
+        transforms.Normalize(
+            mean=[i/255 for i in self.train_mean], 
+            std=[i/255 for i in self.train_stddev])
+    ])
+    transform_config = {
+        "train": transforms.Compose([
+            transforms.RandomResizedCrop(
+                self.center_crop_size,
+                scale=self.scale_ratio_range,
+                ratio=self.aspect_ratio_range),
+            transforms.RandomHorizontalFlip(),
+            normalize]),
+        "test": transforms.Compose([
+            transforms.Resize(self.resize_size),
+            transforms.CenterCrop(self.center_crop_size),
+            normalize])
+    }
+
+    folder = {'train': 'train', 'test': 'val'}
+
+    dataset = ImageFolder(
+        os.path.join(data_dir, folder[split]),
+        transform=transform_config[split])
+
+    dataloader = torch.utils.data.DataLoader(
+        dataset,
+        batch_size=batch_size,
+        shuffle=is_train,
+        num_workers=5,
+        pin_memory=True,
+        # Note (runame): not sure why we do this?
+        drop_last=is_train,
+        # Note (runame): if we want to use fast_collate,
+        # we need to change some things
+        #collate_fn=fast_collate
+    )
+
+    if is_train:
+      dataloader = cycle(dataloader)
+
+    return dataloader
+
+  def preprocess_for_train(
+      self,
+      selected_raw_input_batch: spec.Tensor,
+      selected_label_batch: spec.Tensor,
+      train_mean: spec.Tensor,
+      train_stddev: spec.Tensor,
+      rng: spec.RandomState) -> spec.Tensor:
+    del train_mean
+    del train_stddev
+    del rng
+    return self.preprocess_for_eval(
+        selected_raw_input_batch, selected_label_batch, None, None)
+
+  def preprocess_for_eval(
+      self,
+      raw_input_batch: spec.Tensor,
+      raw_label_batch: spec.Tensor,
+      train_mean: spec.Tensor,
+      train_stddev: spec.Tensor) -> spec.Tensor:
+    del train_mean
+    del train_stddev
+    return (raw_input_batch.float().to(DEVICE), raw_label_batch.to(DEVICE))
+
+  def init_model_fn(self, rng: spec.RandomState) -> spec.ModelInitState:
```

Comment:
What would happen if someone would call the `init_model_fn()` again during training?
We originally envisioned this function to be callable to reset the model parameters. This might be helpful if the model diverges during training.

https://github.com/mlcommons/algorithmic-efficiency/blob/main/RULES.md#model-initialization

Reply:
The model parameters will be reset (using the random seed `rng`), since a new model instance is created. It is handled in the [same way in the MNIST PyTorch workload](https://github.com/mlcommons/algorithmic-efficiency/blob/main/workloads/mnist/mnist_pytorch/workload.py#L113-L119).

However, if people call it during training, they might also need to reinitialize/delete the loss and optimizer (and other objects which might still refer to the old model instance), because the computation graph of the previous instance might be kept around and the memory might not completely be freed, which could lead to OOM errors.

Reply:
Is it correct to say: As long as the submitters are careful, they can reset the model parameters?

Do they have access via our API to properly delete the loss? Handeling the optimizer should be fine.

I remember in the past having a similar problem where I couldn't free memory from an old model in PyTorch. I think this could be a `sample_submission` that we add to our doc in the future. Just a simple optimizer that resets if X happens.

Could you create an issue for this? Either describe how it works and that we should document it and/or provide an example for it. Or - if necessary - that we should investigate this further, whether/how resetting would work.

Reply:
I think the only way to ensure the expected behaviour is to test this properly, maybe through a `sample_submission`, as you have proposed. I have created a corresponding issue (#48).

Reply:
Great! Thanks!